Cython/Compiler/ModuleNode.py

```diff
@@ -2165,6 +2202,52 @@ def generate_property_set_function(self, property_entry, code):
         code.putln(
             "}")
 
+    def generate_typeobj_spec(self, modname, entry, code):
+        type = entry.type
+        scope = type.scope
+        if type.typedef_flag:
+            objstruct = type.objstruct_cname
+        else:
+            objstruct = "struct %s" % type.objstruct_cname
+        classname = scope.class_name.as_c_string_literal()
+        code.putln("static PyType_Slot %s_slots[] = {" % type.typeobj_cname)
+        has_tp_getattro = False
+        for slot in TypeSlots.slot_table:
+            if slot.slot_name == "tp_flags":
+                continue
+            if slot.slot_name == "tp_new" and scope.lookup_here("__cinit__") is None:
```

Comment:
This doesn't look right. `__cinit__` is not the same as `tp_new`. You need to leave the decision to the `TypeSlot` itself, for all slots. It knows when a slot is needed or not.

Seeing this code actually reminded me of an old branch that I had lying around on my hard drive, where I started implementing the same thing (I should get into the habit of creating PRs even from unfinished branches …)
https://github.com/cython/cython/compare/gh2343_heap_types?expand=1

Reply:
Still TODO, will come back to this

Reply:
@scoder this tp_new code is added on purpose. Differently from static types, heap allocated types automatically inherit a `tp_new` function. What I intend to do here is to avoid writing the default `tp_new` that gets added when `__cinit__` is not included. Same for `tp_dealloc`, it now inherits `subtype_dealloc` which know exactly how to dealloc this heap type.

Reply:
I still think it's not correct (at least, `TypeSlots.ConstructorSlot` should be involved in this), but let's leave it at this for now. We'll see into which problems we run later.